docs/source/hooks.rst

```diff
@@ -0,0 +1,136 @@
+* :py:meth:`~flow.FlowProject.add_hook.on_success` will execute when the operation exits without error.
+* :py:meth:`~flow.FlowProject.add_hook.on_fail` will execute when the operation exits with error.
+
+The :py:class:`~flow.FlowProject.add_hook` decorator accepts objects as a function of the job operation
+(:py:class:`~flow.project.JobOperation`).
+The decorators :py:meth:`~flow.FlowProject.add_hook.on_start`, :py:meth:`~flow.FlowProject.add_hook.on_finish`,
+and :py:class:`~flow.FlowProject.add_hook.on_start`
+accept functions with two parameters: the operation name and the :py:class:`Job` object.
+The decorator :py:meth:`~flow.FlowProject.add_hook.on_fail`, accepts functions with three parameters: the operation name, the output error,
+and the :py:class:`Job` object.
+
+:py:class:`~flow.FlowProject.add_hook` can be used to store basic information about the execution of a job operation to the job document.
+
+In the following example, either the function ``store_success_to_doc`` executes after the
+:py:class:`~flow.project.JobOperation`, ``foo``, exits without error, or ``store_error_to_doc`` executes after ``foo``
+exits with error:
+
+.. code-block:: python
+
+    # project.py
+    from flow import FlowProject
+
+    class Project(FlowProject):
+        pass
+
+    def store_success_to_doc(operation_name, job):
+        job.doc.update({f'{operation_name}_success': True})
+
+    def store_error_to_doc(operation_name, error, job):
+        job.doc.update({f'{operation_name}_success': False})
+
+    @FlowProject.operation
+    @FlowProject.add_hook.on_success(store_success_to_doc)
+    @FlowProject.add_hook.on_fail(store_error_to_doc)
+    @FlowProject.post.isfile("result.txt")
+    def foo(job):
+        if job.sp.a == 0:
+            # Have jobs with statepoint 'a' == 0 fail
+            raise ValueError
+
+    if __name__ == '__main__':
+       FlowProject().main()
+
+If ``foo`` is executed using ``python project.py run -o foo -f a 1``, the hook triggered ``on_success`` will run,
+and ``job.doc.get("foo_success") == True``.
+
+If ``foo`` is executed using ``python project.py run -o foo -f a 0``, a ``ValueError`` is raised.
+The hook triggered ``on_fail`` will run, and ``job.doc.get("foo_success") == False``.
+
+.. note::
+
+    Unlike :py:meth:`~flow.FlowProject.add_hook.on_start`, :py:meth:`~flow.FlowProject.add_hook.on_finish`,
+    and :py:meth:`~flow.FlowProject.add_hook.on_on_success`,
+    which accept functions that take ``operation_name`` and ``job`` as arguments,
+    :py:meth:`~flow.FlowProject.add_hook.on_fail` accepts functions that take ``operation_name``, ``error``,
+    and ``job`` as arguments.
+
+.. _project-level hooks:
+
+Project-Level Hooks
+===================
+
+In some cases, it may be desirable to install the same hook or set of hooks for all operations in a project.
+A custom set of hooks may be installed by a custom ``install_hooks`` method:
```

Comment:
it's not clear to me where a ``install_hooks`` method is here

Reply:
Ah that's because there isn't one. I couldn't think of a simple, straight forward example that creates a hook system in a class. I changed it so it reflects the text, but it feels a bit contrived. What do you think?